(React SDK): Forward refs in withOptimizely HoC

mjc1283 opened the pull request:
Implement ref forwarding in `withOptimizely` HoC.

jamesopti commented on packages/react-sdk/src/withOptimizely.tsx:
```diff
@@ -17,36 +17,44 @@ import * as React from 'react'
+      </OptimizelyContextConsumer>
+    )
   }
+  withOptimizely.displayName = `withOptimizely(${Component.displayName ||
```
Checkout the helper I wrote in the segment-js repo to forward refs and hoist non-react statics.

I think copy/pasting that here would be useful.

https://github.com/optimizely/segment-js/blob/master/src/decorators.js#L14-L29

mjc1283 replied:
@jamesopti I tried using this helper function, but I'm finding this to make the types a lot more challenging. Would you be OK if we defer this until later? FWIW, this `withOptimizely` is the only HoC that would be using this helper function at the moment.

jamesopti commented on packages/react-sdk/src/withOptimizely.spec.tsx:
```diff
@@ -192,4 +192,35 @@ describe('withOptimizely', () => {
+      </OptimizelyProvider>,
+    )
+    expect(inputRef.current).toBeInstanceOf(HTMLInputElement)
+    expect(typeof inputRef.current!.focus).toBe('function')
```
Can you also assert that `inputRef.current.value` is in fact the same reference as `component.find('input'). getDOMNode()`?

mjc1283 commented:
@jamesopti @mikeng13 I refactored this a little and combined it with hoisting statics. Moved into a separate PR: https://github.com/optimizely/fullstack-labs/pull/32
